# Pandas - DataFrame

### DataFrame

É um objeto que representa uma matriz de 2 dimensões, onde todas as colunas são alinhadas pelo mesmo índice. Um dataframe pode ter séries (colunas) com tipos diferentes.

Pode ser comparado a uma planilha excel ou a uma tabela de banco de dados.


![DataFrame](imagens/excel_dataframe.png)

In [1]:
import warnings
warnings.filterwarnings('ignore')

# por convenção utiliza-se o alias np ao importar a biblioteca
import numpy as np

# por convenção utiliza-se o alias pd ao importar a biblioteca
import pandas as pd

import matplotlib.pyplot as plt
%matplotlib inline

# outros imports que serão úteis
import os
import re
from collections import OrderedDict

# forçar a renderização do dataframe sem notação científica
pd.set_option('display.float_format', lambda x: '{:.2f}'.format(x))

In [4]:
s5 = pd.Series([27, 28.3, 29.1], 
               name='Temperatura Sensor A', 
               index=['2018-02-01', '2018-02-02', '2018-02-03'])
s6 = pd.Series([28.6, 32.1, 26], 
               name='Temperatura Sensor B', 
               index=['2018-02-02', '2018-02-03', '2018-02-04'])
s7 = (s5 + s6) / 2
s7 = s7.rename('Temperatura Média')

# Criação de um DataFrame à partir de um dicionário de séries
d = {
    'Sensor A': s5,
    'Sensor B': s6,
    'Médias': s7,
}
df1 = pd.DataFrame(d)
df1

Sensor A  Sensor B  Médias
2018-02-01     27.00       nan     nan
2018-02-02     28.30     28.60   28.45
2018-02-03     29.10     32.10   30.60
2018-02-04       nan     26.00     nan

In [3]:
# se passarmos uma lista com os índices, todos os dados que não estiverem
# de acordo com os índices informados serão descartados
df2 = pd.DataFrame(d, index=['2018-02-02', '2018-02-03'])
df2

Sensor A  Sensor B  Médias
2018-02-02     28.30     28.60   28.45
2018-02-03     29.10     32.10   30.60

In [7]:
# Criação de um DataFrame à partir de um dicionário de listas

d = {
    'Produto': ['Maçã', 'Leite', 'Canela'],
    'Unidade de Medida': ['kg', 'L', 'g'],
    'Valor Unitário': [4.99, 6.7, 1.6],
}
df3 = pd.DataFrame(d)
df3

Produto Unidade de Medida  Valor Unitário
0    Maçã                kg            4.99
1   Leite                 L            6.70
2  Canela                 g            1.60

In [8]:
# Se nenhum índice é informado, o DataFrame terá um RangeIndex iniciando em 0.
print(df3.index)

RangeIndex(start=0, stop=3, step=1)


In [6]:
# Criação de um DataFrame à partir de um dicionário de listas

d = dict({
    'Produto': ['Maçã', 'Leite', 'Canela'],
    'Unidade de Medida': ['kg', 'L', 'g'],
    'Valor Unitário': [4.99, 6.7, 1.6],
})
df3 = pd.DataFrame(d)
df3

Produto Unidade de Medida  Valor Unitário
0    Maçã                kg            4.99
1   Leite                 L            6.70
2  Canela                 g            1.60

In [10]:
# Uma série pode ser transformada em um DataFrame de uma coluna utilizando o método 'to_frame'
df4 = s7.to_frame()
df4


Temperatura Média
2018-02-01                nan
2018-02-02              28.45
2018-02-03              30.60
2018-02-04                nan

* Características e atributos importantes do dataframes


In [8]:
print('Valores:', df1.values, '\nTipo:', type(df1.values))
print('\nÍndices:', df1.index, '\nTipo:', type(df1.index))
print('\nNome:', df1.columns, '\nTipo:', type(df1.columns))
print('\nShape:', df1.shape, '\nTipo:', type(df1.shape))

Valores: [[27.     nan   nan]
 [28.3  28.6  28.45]
 [29.1  32.1  30.6 ]
 [  nan 26.     nan]] 
Tipo: <class 'numpy.ndarray'>

Índices: Index(['2018-02-01', '2018-02-02', '2018-02-03', '2018-02-04'], dtype='object') 
Tipo: <class 'pandas.core.indexes.base.Index'>

Nome: Index(['Sensor A', 'Sensor B', 'Médias'], dtype='object') 
Tipo: <class 'pandas.core.indexes.base.Index'>

Shape: (4, 3) 
Tipo: <class 'tuple'>


O dataframe possui o método info() que descreve algumas características básicas como, tipo de dado, quantidade de registros, intervalo do índice, registros não nulos por coluna e tamanho em memória.

In [9]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, 2018-02-01 to 2018-02-04
Data columns (total 3 columns):
Sensor A    3 non-null float64
Sensor B    3 non-null float64
Médias      2 non-null float64
dtypes: float64(3)
memory usage: 128.0+ bytes


A operação padrão de indexação/fatiamento do DataFrame é um pouco diferente da da Série. Enquanto a série representa apenas uma coluna, o dataframe pode conter várias colunas.

Dessa forma, os slices no dataframe podem abranger linhas e/ou colunas. Abaixo veremos uma série de exemplos simples de slices em DataFrames. Seleções mais complexas e completas serão abordadas em uma seção específica no próximo capítulo.

In [10]:
# seleção de uma coluna no dataframe retorna uma série
df1['Sensor A']

2018-02-01   27.00
2018-02-02   28.30
2018-02-03   29.10
2018-02-04     nan
Name: Sensor A, dtype: float64

In [11]:
df1[['Sensor A']]

Sensor A
2018-02-01     27.00
2018-02-02     28.30
2018-02-03     29.10
2018-02-04       nan

In [11]:
# seleção de mais de uma coluna no dataframe retorna um novo dataframe com a seleção
df1[['Sensor A', 'Sensor B']]

Sensor A  Sensor B
2018-02-01     27.00       nan
2018-02-02     28.30     28.60
2018-02-03     29.10     32.10
2018-02-04       nan     26.00

In [12]:
# se for especificada uma ordem diferente um novo dataframe sera 
#criado com as colunas na ordem informada
df1[['Sensor B', 'Sensor A']]

Sensor B  Sensor A
2018-02-01       nan     27.00
2018-02-02     28.60     28.30
2018-02-03     32.10     29.10
2018-02-04     26.00       nan

In [13]:
# um dataframe pode ser criado à partir de outro de forma fácil
df5 = df1.copy()

#ou

df5 = pd.DataFrame(df1)

df5

Sensor A  Sensor B  Médias
2018-02-01     27.00       nan     nan
2018-02-02     28.30     28.60   28.45
2018-02-03     29.10     32.10   30.60
2018-02-04       nan     26.00     nan

In [14]:
#novas colunas podem ser criadas no dataframe à partir de um valor escalar
df5['Sensor C'] = 30
df5

Sensor A  Sensor B  Médias  Sensor C
2018-02-01     27.00       nan     nan        30
2018-02-02     28.30     28.60   28.45        30
2018-02-03     29.10     32.10   30.60        30
2018-02-04       nan     26.00     nan        30

In [15]:
# novas colunas podem ser criadas no dataframe à partir de listas,
# desde que o número de linhas seja igual ao do dataframe
df5['Sensor C'] = [1,2,3,4]
df5

Sensor A  Sensor B  Médias  Sensor C
2018-02-01     27.00       nan     nan         1
2018-02-02     28.30     28.60   28.45         2
2018-02-03     29.10     32.10   30.60         3
2018-02-04       nan     26.00     nan         4

In [16]:
# colunas podem ser removidas com o comando del (veja na mais sobre o del na documentação do python)
del(df5['Sensor C'])
del(df5['Médias'])
df5

Sensor A  Sensor B
2018-02-01     27.00       nan
2018-02-02     28.30     28.60
2018-02-03     29.10     32.10
2018-02-04       nan     26.00

In [18]:
# os índices podem ter nomes atribuídos
df5.columns.name = 'Sensores'
df5.index.name = 'Data'

df5

Sensores    Sensor A  Sensor B
Data                          
2018-02-01     27.00       nan
2018-02-02     28.30     28.60
2018-02-03     29.10     32.10
2018-02-04       nan     26.00

A maioria dos métodos vistos para Series funcionam para DataFrames também, mas podem ter variações ou limitações dependendo do tipo de dado de cada série.

In [19]:
print('min:\n', df1.min())

# maior valor
print('\nmax:\n', df1.max())

# média
print('\nmean:\n', df1.mean())

# mediana
print('\nmedian:\n', df1.median())

min:
 Sensores
Sensor A   27.00
Sensor B   26.00
dtype: float64

max:
 Sensores
Sensor A   29.10
Sensor B   32.10
dtype: float64

mean:
 Sensores
Sensor A   28.13
Sensor B   28.90
dtype: float64

median:
 Sensores
Sensor A   28.30
Sensor B   28.60
dtype: float64


In [26]:
df_tmp = pd.DataFrame([[2,2,2,2,2],
              [3,3,3,3,3],
              [4,4,4,4,4],
              [5,5,5,5,5],
              [6,6,6,6,6],
             ], columns=['A','B','C','D','E'], index=[1,2,3,4,5])
df_tmp['F'] = pd.Series([7,7,7,7,7], index=[5,7,8,9,10])
df_tmp[df_tmp['E'].isin([4,3])].reset_index(drop=True)

A  B  C  D  E   F
0  3  3  3  3  3 nan
1  4  4  4  4  4 nan

Documentação do Pandas para a classe DataFrame:
https://pandas.pydata.org/pandas-docs/stable/api.html#series

Outras formas de criar DataFrames podem ser consultadas aqui: http://pandas.pydata.org/pandas-docs/stable/dsintro.html#alternate-constructors

## Seleções e subconjuntos de dados

Esta seção pretende abordar as  mais variadas formas de seleções de dados utilizando DataFrames, Séries e Índices.

### Expressões

Expressões são operações que podem ser feitas com DataFrames, Séries e Índices utilizando operadores lógicos e de comparação, com o objetivo de criar filtros para gerar subconjuntos de dados.

In [185]:
def to_float(valor):
    if valor == 'ni':
        return np.nan
    return float(valor.replace('.','').replace(',','.').replace(' ',''))
    
df = pd.read_csv(
    'dados/arq_municipios_fronteiricos.tsv', 
    sep='\t',
    decimal='.', 
    thousands=',', 
    converters={
        'Área territorial': to_float,
        'População (IBGE/2007)': to_float,
        'Densidade demográfica (hab/km2)': to_float,
        'PIB (IBGE/2005)': to_float,
        'PIB per capita (R$)': to_float,
        'IDH/2000': to_float,
    }, 
    na_values=['ni'],
)

df['Município'] = df['Município'].str.replace(r'([\d]* [-] )(.*)', r'\2')

print(df.shape)
df.head()

(122, 8)


Município             Estado  Área territorial  \
0                 Aceguá  Rio Grande do Sul           1550.00   
1             Acrelândia               Acre           1575.00   
2                Alecrim  Rio Grande do Sul            315.00   
3               Almeirim               Pará          72960.00   
4  Alta Floresta d'Oeste           Rondônia           7067.00   

   População (IBGE/2007)  Densidade demográfica (hab/km2)  PIB (IBGE/2005)  \
0                4138.00                             2.66      71638000.00   
1               11520.00                             7.31     114350000.00   
2                7357.00                            23.35      44373000.00   
3               30903.00                             0.42     462258000.00   
4               23857.00                             3.37     186812000.00   

   PIB per capita (R$)  IDH/2000  
0             17266.00       nan  
1              9986.00      0.68  
2              5944.00      0.74  
3             13485.00      0.74  
4              6525.00      0.71

Vejamos um exemplo de expressão:

In [186]:
df[['IDH/2000']].head()

IDH/2000
0       nan
1      0.68
2      0.74
3      0.74
4      0.71

In [187]:
# utilizando um operador de comparação associado a uma 
# estrutura de dados do pandas temos uma expressão
expressao = df['IDH/2000'] >= 0.741

print(type(expressao))

expressao.head()

<class 'pandas.core.series.Series'>


0    False
1    False
2     True
3     True
4    False
Name: IDH/2000, dtype: bool

In [188]:
df['IDH Médio ou Alto'] = expressao.apply(lambda x: 1 if x else 0)
df.head()

Município             Estado  Área territorial  \
0                 Aceguá  Rio Grande do Sul           1550.00   
1             Acrelândia               Acre           1575.00   
2                Alecrim  Rio Grande do Sul            315.00   
3               Almeirim               Pará          72960.00   
4  Alta Floresta d'Oeste           Rondônia           7067.00   

   População (IBGE/2007)  Densidade demográfica (hab/km2)  PIB (IBGE/2005)  \
0                4138.00                             2.66      71638000.00   
1               11520.00                             7.31     114350000.00   
2                7357.00                            23.35      44373000.00   
3               30903.00                             0.42     462258000.00   
4               23857.00                             3.37     186812000.00   

   PIB per capita (R$)  IDH/2000  IDH Médio ou Alto  
0             17266.00       nan                  0  
1              9986.00      0.68                  0  
2              5944.00      0.74                  1  
3             13485.00      0.74                  1  
4              6525.00      0.71                  0

In [189]:
df.head()

Município             Estado  Área territorial  \
0                 Aceguá  Rio Grande do Sul           1550.00   
1             Acrelândia               Acre           1575.00   
2                Alecrim  Rio Grande do Sul            315.00   
3               Almeirim               Pará          72960.00   
4  Alta Floresta d'Oeste           Rondônia           7067.00   

   População (IBGE/2007)  Densidade demográfica (hab/km2)  PIB (IBGE/2005)  \
0                4138.00                             2.66      71638000.00   
1               11520.00                             7.31     114350000.00   
2                7357.00                            23.35      44373000.00   
3               30903.00                             0.42     462258000.00   
4               23857.00                             3.37     186812000.00   

   PIB per capita (R$)  IDH/2000  IDH Médio ou Alto  
0             17266.00       nan                  0  
1              9986.00      0.68                  0  
2              5944.00      0.74                  1  
3             13485.00      0.74                  1  
4              6525.00      0.71                  0

In [190]:
# É possível utilizar o apply no DataFrame para executar operações em mais de uma coluna
df.apply(lambda linha: linha['População (IBGE/2007)']/linha['Área territorial'], axis=1)

0      2.67
1      7.31
2     23.36
3      0.42
4      3.38
5      0.56
6      2.93
7      0.27
8      7.30
9      5.58
10     1.86
11     0.18
12    27.48
13    20.74
14     0.20
15     3.58
16    55.04
17     4.67
18    28.52
19     1.05
20    22.04
21     1.26
22     4.40
23     5.00
24     3.45
25    43.21
26     4.93
27     0.38
28     1.73
29     0.59
       ... 
92    38.64
93    25.71
94    30.28
95     7.16
96     3.76
97    21.03
98    30.07
99    30.09
100    0.27
101    0.66
102    5.95
103   12.01
104    2.38
105   56.95
106   18.16
107    3.31
108    0.36
109   48.92
110   29.78
111   12.18
112    1.35
113    8.94
114   12.90
115   14.04
116   29.61
117   34.96
118    0.92
119   21.65
120    1.02
121    2.73
Length: 122, dtype: float64

In [42]:
expressao.describe()

count       122
unique        2
top       False
freq         70
Name: IDH/2000, dtype: object

As expressões são séries ou dataframes com o mesmo tamanho dos originais, porém seus valores são preenchidos com True ou False de acordo com a avaliação da expressão para cada item. Essas séries ou dataframes de booleanos podem ser passados para os operadores de indexação __[ ]__ e __loc__.

In [45]:
print(df[expressao].shape)

df[expressao].head()

# ou 

df.loc[expressao].head()

(52, 9)


Município             Estado  Área territorial  \
2           Alecrim  Rio Grande do Sul            315.00   
3          Almeirim               Pará          72960.00   
12             Bagé  Rio Grande do Sul           4096.00   
13      Bandeirante     Santa Catarina            146.00   
15  Barra do Quaraí  Rio Grande do Sul           1056.00   

    População (IBGE/2007)  Densidade demográfica (hab/km2)  PIB (IBGE/2005)  \
2                 7357.00                            23.35      44373000.00   
3                30903.00                             0.42     462258000.00   
12              112550.00                            27.47     906488000.00   
13                3028.00                            20.73      21423000.00   
15                3776.00                             3.57      61540000.00   

    PIB per capita (R$)  IDH/2000  IDH Médio ou Alto  
2               5944.00      0.74                  1  
3              13485.00      0.74                  1  
12              7473.00      0.80                  1  
13              7546.00      0.77                  1  
15             14429.00      0.78                  1

Podemos utilizar operadores lógicos para avaliar mais de uma expressão caso sejam necessárias condicionais mais complexas para os filtros.

In [47]:
a = 1
b = 2
c = 3
if a == 1 and b == 2:
    print('Entrou no if')

Entrou no if


In [46]:
expressao = (df['IDH/2000'] >= .741) & (df['Estado'] == 'Rio Grande do Sul')
print(df[expressao].shape)
df.loc[expressao].head()

(23, 9)


Município             Estado  Área territorial  \
2           Alecrim  Rio Grande do Sul            315.00   
12             Bagé  Rio Grande do Sul           4096.00   
15  Barra do Quaraí  Rio Grande do Sul           1056.00   
30             Chuí  Rio Grande do Sul            203.00   
35       Crissiumal  Rio Grande do Sul            362.00   

    População (IBGE/2007)  Densidade demográfica (hab/km2)  PIB (IBGE/2005)  \
2                 7357.00                            23.35      44373000.00   
12              112550.00                            27.47     906488000.00   
15                3776.00                             3.57      61540000.00   
30                5278.00                            26.00      67525000.00   
35               14726.00                            40.67     114089000.00   

    PIB per capita (R$)  IDH/2000  IDH Médio ou Alto  
2               5944.00      0.74                  1  
12              7473.00      0.80                  1  
15             14429.00      0.78                  1  
30             10574.00      0.81                  1  
35              8376.00      0.79                  1

In [57]:
lista_idx = list(df[df['Estado'] == 'Acre'].index)

df.loc[lista_idx]

Município Estado  Área territorial  População (IBGE/2007)  \
1              Acrelândia   Acre           1575.00               11520.00   
10           Assis Brasil   Acre           2876.00                5351.00   
22              Brasiléia   Acre           4336.00               19065.00   
26               Capixaba   Acre           1713.00                8446.00   
36        Cruzeiro do Sul   Acre           7925.00               73948.00   
42         Epitaciolândia   Acre           1659.00               13434.00   
44                  Feijó   Acre          24202.00               31288.00   
59                 Jordão   Acre           5429.00                6059.00   
61            Mâncio Lima   Acre           4672.00               13785.00   
62          Manoel Urbano   Acre           9387.00                7148.00   
64   Marechal Thaumaturgo   Acre           7744.00               13061.00   
81      Plácido de Castro   Acre           2047.00               17258.00   
91           Porto Walter   Acre           6136.00                8170.00   
96        Rodrigues Alves   Acre           3305.00               12428.00   
101   Santa Rosa do Purus   Acre           5981.00                3948.00   
112        Sena Madureira   Acre          25278.00               34230.00   
121                Xapuri   Acre           5251.00               14314.00   

     Densidade demográfica (hab/km2)  PIB (IBGE/2005)  PIB per capita (R$)  \
1                               7.31     114350000.00              9986.00   
10                              1.86      30298000.00              5984.00   
22                              4.39     117525000.00              6632.00   
26                              4.93      78389000.00             11092.00   
36                              9.33     391943000.00              4647.00   
42                              8.59      88164000.00              6397.00   
44                              1.29     144977000.00              3791.00   
59                              1.11      23805000.00              5138.00   
61                              2.95      49705000.00              3899.00   
62                              0.76      31268000.00              4095.00   
64                              1.68      37645000.00              4452.00   
81                              8.43     139814000.00              8377.00   
91                              1.33      25925000.00              5225.00   
96                              3.76      51468000.00              5254.00   
101                             0.66      15321000.00              4513.00   
112                             1.35     234381000.00              7105.00   
121                             2.72      82377000.00              6016.00   

     IDH/2000  IDH Médio ou Alto  
1        0.68                  0  
10       0.67                  0  
22       0.67                  0  
26       0.61                  0  
36       0.67                  0  
42       0.68                  0  
44       0.54                  0  
59       0.47                  0  
61       0.64                  0  
62       0.60                  0  
64       0.53                  0  
81       0.68                  0  
91       0.54                  0  
96       0.55                  0  
101      0.53                  0  
112      0.65                  0  
121      0.67                  0

Os comparadores lógicos são:
* __<__  : menor que ...
* __>__  : maior que ...
* __==__  : igual a  ...
* __<=__  : menor ou igual a  ...
* __>=__  : maior ou igual a  ...
* __!=__  : diferente de ...
* __isnull()__  : se o valor é nulo ou inválido (NaN) ...
* __notnull()__  : se o valor não é nulo e nem inválido (NaN) ...
* __isin()__  : se o valor está contido em um dos dados informados em uma lista ...

Os operadores lógicos são:
* __&__: operador E (and)
* __|__: operador OU (or)
* __~__: operador NAO (not)
* __^__: operador OU EXCLUSIVO (xor)
* __any()__: se algum elemento da lista for True retornará True
* __all()__: se todos os elementos da lista forem True retornará True

### Opções de indexação

__df[val]__
* seleciona uma ou uma sequência de colunas em um DataFrame
* seleciona linhas com expressões
* seleciona linhas por posição
* seleciona linhas por índice (label), cuidado, resultado pode ser ambíguo (utilize loc neste caso)
* caso esteja sendo feito algum filtro de linhas, não permite seleção de colunas ao mesmo tempo. Retornará todas
 
__df.loc[val]__
* seleciona uma ou mais linhas pelo índice

__df.loc[:, val]__
* seleciona uma ou mais colunas pelo nome das colunas

__df.loc[val1, val2]__
* seleciona linhas e colunas

__df.iloc[onde]__
* seleciona uma ou mais linhas pela posição. 'onde' pode ser uma posição, slice ou uma lista de posições

__df.iloc[:, onde]__
* seleciona uma ou mais colunas pela posição. 'onde' pode ser uma posição, slice ou uma lista de posições

__df.iloc[onde_i, onde_j]__
* seleciona uma ou mais linhas e colunas pela posição. 'onde_?' pode ser uma posição, slice ou uma lista de posições

# Exercício

Para todos os items abaixo, considere o dataframe contido na variável "df".

1. Utilize os métodos head, tail e sample para visualizar os dados do DataFrame.
2. Imprima o número de linhas e o número de colunas deste DataFrame.
3. Imprima o valor do maior PIB
4. Imprima a linha do município que contém o maior PIB
5. Ordene o dataframe pelo PIB em ordem decrescente
6. Faça uma seleção com as colunas Município, Estado e IDH, atribua à variável df_idh e imprima as primeiras 3 linhas
7. Faça uma seleção com as colunas Município, Estado e Densidade demográfica (hab/km2), e com as linhas de índices 1, 2 e 3. Utilize o método loc.
8. Filtre apenas os municípios do Pará
9. Filtre apenas os municípios do Acre com IDH menor que 0.6
10. Encontre os municípios entre os 25% com maior PIB (IBGE/2005) e faça a contagem desses municípios por estado (Dicas: quantile e value_counts).

In [69]:

# Utilize os métodos head, tail e sample para visualizar os dados do DataFrame.
from IPython.display import display, Markdown

display(df.head())
display(df.tail())
display(df.sample())

Município             Estado  Área territorial  \
0                 Aceguá  Rio Grande do Sul           1550.00   
1             Acrelândia               Acre           1575.00   
2                Alecrim  Rio Grande do Sul            315.00   
3               Almeirim               Pará          72960.00   
4  Alta Floresta d'Oeste           Rondônia           7067.00   

   População (IBGE/2007)  Densidade demográfica (hab/km2)  PIB (IBGE/2005)  \
0                4138.00                             2.66      71638000.00   
1               11520.00                             7.31     114350000.00   
2                7357.00                            23.35      44373000.00   
3               30903.00                             0.42     462258000.00   
4               23857.00                             3.37     186812000.00   

   PIB per capita (R$)  IDH/2000  
0             17266.00       nan  
1              9986.00      0.68  
2              5944.00      0.74  
3             13485.00      0.74  
4              6525.00      0.71

Município             Estado  Área territorial  \
117                         Tunápolis      Santa Cataria            133.00   
118                          Uiramutã            Roraima           8066.00   
119                        Uruguaiana  Rio Grande do Sul           5716.00   
120  Vila Bela da Santíssima Trindade        Mato Grosso          13631.00   
121                            Xapuri               Acre           5251.00   

     População (IBGE/2007)  Densidade demográfica (hab/km2)  PIB (IBGE/2005)  \
117                4650.00                            34.96      40845000.00   
118                7403.00                             0.91      27251000.00   
119              123743.00                            21.64    1187038000.00   
120               13886.00                             1.01     116908000.00   
121               14314.00                             2.72      82377000.00   

     PIB per capita (R$)  IDH/2000  
117              9420.00      0.82  
118              4238.00      0.54  
119              8798.00      0.79  
120              8047.00      0.71  
121              6016.00      0.67

Município             Estado  Área territorial  \
116  Tiradentes do Sul  Rio Grande do Sul            234.00   

     População (IBGE/2007)  Densidade demográfica (hab/km2)  PIB (IBGE/2005)  \
116                6928.00                            29.60      42926000.00   

     PIB per capita (R$)  IDH/2000  
116              6577.00      0.75

In [70]:
# Imprima o número de linhas e o número de colunas deste DataFrame.
df.shape

(122, 8)

In [71]:
# Imprima o valor do maior PIB
df['PIB (IBGE/2005)'].max()

4853331000.0

In [77]:
# Imprima a linha do município que contém o maior PIB
display(df[df['PIB (IBGE/2005)'] == df['PIB (IBGE/2005)'].max()])

id=df['PIB (IBGE/2005)'].idxmax()
display(id)
display(df.loc[[id]])

display(df.sort_values('PIB (IBGE/2005)', ascending=False).head(1))

display(df.sort_values('PIB (IBGE/2005)', ascending=True).tail(1))

Município  Estado  Área territorial  População (IBGE/2007)  \
45  Foz do Iguaçu  Paraná            618.00              311336.00   

    Densidade demográfica (hab/km2)  PIB (IBGE/2005)  PIB per capita (R$)  \
45                           503.70    4853331000.00             16102.00   

    IDH/2000  
45      0.80

45

Município  Estado  Área territorial  População (IBGE/2007)  \
45  Foz do Iguaçu  Paraná            618.00              311336.00   

    Densidade demográfica (hab/km2)  PIB (IBGE/2005)  PIB per capita (R$)  \
45                           503.70    4853331000.00             16102.00   

    IDH/2000  
45      0.80

Município  Estado  Área territorial  População (IBGE/2007)  \
45  Foz do Iguaçu  Paraná            618.00              311336.00   

    Densidade demográfica (hab/km2)  PIB (IBGE/2005)  PIB per capita (R$)  \
45                           503.70    4853331000.00             16102.00   

    IDH/2000  
45      0.80

Município  Estado  Área territorial  População (IBGE/2007)  \
45  Foz do Iguaçu  Paraná            618.00              311336.00   

    Densidade demográfica (hab/km2)  PIB (IBGE/2005)  PIB per capita (R$)  \
45                           503.70    4853331000.00             16102.00   

    IDH/2000  
45      0.80

In [78]:
# Ordene o dataframe pelo PIB em ordem decrescente
df.sort_values('PIB (IBGE/2005)', ascending=False).head()

Município              Estado  Área territorial  \
45   Foz do Iguaçu              Paraná            618.00   
89     Porto Velho            Rondônia          34082.00   
33         Corumbá  Mato Grosso do Sul          65304.00   
119     Uruguaiana   Rio Grande do Sul           5716.00   
12            Bagé   Rio Grande do Sul           4096.00   

     População (IBGE/2007)  Densidade demográfica (hab/km2)  PIB (IBGE/2005)  \
45               311336.00                           503.70    4853331000.00   
89               369345.00                            10.83    3656512000.00   
33               114279.00                             1.74    1557253000.00   
119              123743.00                            21.64    1187038000.00   
12               112550.00                            27.47     906488000.00   

     PIB per capita (R$)  IDH/2000  
45              16102.00      0.80  
89               9779.00      0.76  
33              13234.00      0.77  
119              8798.00      0.79  
12               7473.00      0.80

In [80]:
# Faça uma seleção com as colunas Município, Estado e IDH, atribua à variável df_idh e imprima as primeiras 3 linhas
df_idh = df[['Município', 'Estado','IDH/2000']]
df_idh.head()

Município             Estado  IDH/2000
0                 Aceguá  Rio Grande do Sul       nan
1             Acrelândia               Acre      0.68
2                Alecrim  Rio Grande do Sul      0.74
3               Almeirim               Pará      0.74
4  Alta Floresta d'Oeste           Rondônia      0.71

In [83]:
# Faça uma seleção com as colunas Município, Estado e Densidade demográfica (hab/km2), 
# e com as linhas de índices 1, 2 e 3. Utilize o método loc.
display(df.loc[[1,2,3], ['Município', 'Estado','Densidade demográfica (hab/km2)']])

display(df.loc[1:3, ['Município', 'Estado','Densidade demográfica (hab/km2)']])


Município             Estado  Densidade demográfica (hab/km2)
1  Acrelândia               Acre                             7.31
2     Alecrim  Rio Grande do Sul                            23.35
3    Almeirim               Pará                             0.42

Município             Estado  Densidade demográfica (hab/km2)
1  Acrelândia               Acre                             7.31
2     Alecrim  Rio Grande do Sul                            23.35
3    Almeirim               Pará                             0.42

In [85]:
# Filtre apenas os municípios do Pará
df[ df['Estado'] == 'Pará' ]

Município Estado  Área territorial  População (IBGE/2007)  \
3    Almeirim   Pará          72960.00               30903.00   
70     Óbidos   Pará          28021.00               46793.00   
72  Oriximiná   Pará         107603.00               55175.00   

    Densidade demográfica (hab/km2)  PIB (IBGE/2005)  PIB per capita (R$)  \
3                              0.42     462258000.00             13485.00   
70                             1.66     138489000.00              2820.00   
72                             0.51     776845000.00             14620.00   

    IDH/2000  
3       0.74  
70      0.68  
72      0.72

In [87]:
# Filtre apenas os municípios do Acre com IDH menor que 0.6
df[ (df['Estado'] == 'Acre') & (df['IDH/2000'] < .6) ]

Município Estado  Área territorial  População (IBGE/2007)  \
44                  Feijó   Acre          24202.00               31288.00   
59                 Jordão   Acre           5429.00                6059.00   
64   Marechal Thaumaturgo   Acre           7744.00               13061.00   
91           Porto Walter   Acre           6136.00                8170.00   
96        Rodrigues Alves   Acre           3305.00               12428.00   
101   Santa Rosa do Purus   Acre           5981.00                3948.00   

     Densidade demográfica (hab/km2)  PIB (IBGE/2005)  PIB per capita (R$)  \
44                              1.29     144977000.00              3791.00   
59                              1.11      23805000.00              5138.00   
64                              1.68      37645000.00              4452.00   
91                              1.33      25925000.00              5225.00   
96                              3.76      51468000.00              5254.00   
101                             0.66      15321000.00              4513.00   

     IDH/2000  
44       0.54  
59       0.47  
64       0.53  
91       0.54  
96       0.55  
101      0.53

In [106]:
display(list(df['PIB (IBGE/2005)'].quantile([.75]))[0])
display(df['PIB (IBGE/2005)'].quantile([.75])[0.75])
display(df['PIB (IBGE/2005)'].quantile([.75]).loc[0.75])
display(df['PIB (IBGE/2005)'].quantile([.75]).iloc[0])

151535750.0

151535750.0

151535750.0

151535750.0

In [110]:
# Encontre os municípios entre os 25% com maior PIB (IBGE/2005) e 
# faça a contagem desses municípios por estado (Dicas: quantile e value_counts).
pib = df['PIB (IBGE/2005)'].quantile([.75]).iloc[0]
df[ df['PIB (IBGE/2005)'] > pib ]['Estado'].value_counts()

Rio Grande do Sul     9
Paraná                6
Rondônia              3
Mato Grosso           3
Santa Catarina        3
Pará                  2
Mato Grosso do Sul    2
Acre                  2
Amapá                 1
Name: Estado, dtype: int64

# Agregações

In [112]:
df = pd.read_parquet('dados/bi_ativ_urb_cid.parquet')
for c in ['Espécie', 'Capitulo da CID', 'Sexo', 'Clientela', 'Grupo/Principais Espécies']:
    df[c] = df[c].astype('category')
print(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 547711 entries, 0 to 547710
Data columns (total 9 columns):
Ano                             547711 non-null float64
Espécie                         547711 non-null category
Capitulo da CID                 547711 non-null category
Sexo                            547711 non-null category
Clientela                       547711 non-null category
Quantidade Benefícios Ativos    547711 non-null float64
Valor Benefícios Ativos (R$)    547711 non-null float64
Grupo/Principais Espécies       547711 non-null category
Valor (R$ mil)                  547711 non-null float64
dtypes: category(5), float64(4)
memory usage: 23.5 MB
None


Ano                    Espécie  \
0 2004.00  Ap Invalidez Det Ignorado   
1 2004.00  Ap Invalidez Det Ignorado   
2 2004.00  Ap Invalidez Det Ignorado   
3 2004.00  Ap Invalidez Det Ignorado   
4 2004.00  Ap Invalidez Det Ignorado   

                                 Capitulo da CID       Sexo Clientela  \
0  Cid 9 - I - Doenças Infecciosas e Parasitária  Masculino    Urbana   
1  Cid 9 - I - Doenças Infecciosas e Parasitária   Feminino    Urbana   
2  Cid 9 - I - Doenças Infecciosas e Parasitária   Ignorado    Urbana   
3                        Cid 9 - Ii - Neoplasmas  Masculino    Urbana   
4                        Cid 9 - Ii - Neoplasmas   Feminino    Urbana   

   Quantidade Benefícios Ativos  Valor Benefícios Ativos (R$)  \
0                      18646.00                    5795330.45   
1                       9765.00                    2343602.54   
2                          5.00                       1093.15   
3                       8741.00                    3935448.30   
4                      12185.00                    3661307.37   

  Grupo/Principais Espécies  Valor (R$ mil)  
0              Ap Invalidez         5795.33  
1              Ap Invalidez         2343.60  
2              Ap Invalidez            1.09  
3              Ap Invalidez         3935.45  
4              Ap Invalidez         3661.31

In [116]:
# Agregação simples com uso de somatório

df_ano_sexo = df[['Ano', 'Sexo', 'Quantidade Benefícios Ativos', 'Valor Benefícios Ativos (R$)']]\
    .groupby(['Ano', 'Sexo'], as_index=False).sum()
df_ano_sexo.head(6)

Ano       Sexo  Quantidade Benefícios Ativos  \
0 2004.00   Feminino                    1396298.00   
1 2004.00   Ignorado                      29175.00   
2 2004.00  Masculino                    1959787.00   
3 2005.00   Feminino                    1460602.00   
4 2005.00   Ignorado                      25510.00   
5 2005.00  Masculino                    2041717.00   

   Valor Benefícios Ativos (R$)  
0                  550962123.88  
1                    7708856.49  
2                 1131942614.36  
3                  629946761.14  
4                    7185139.31  
5                 1299889240.05

In [121]:
grupos = df[['Ano', 'Sexo', 'Quantidade Benefícios Ativos', 'Valor Benefícios Ativos (R$)']]\
    .groupby(['Ano', 'Sexo'], as_index=False).agg({
        'Quantidade Benefícios Ativos':'sum',
        'Valor Benefícios Ativos (R$)':'mean',
})
grupos[ grupos['Valor Benefícios Ativos (R$)'] >= 3443513]

Ano       Sexo  Quantidade Benefícios Ativos  \
0 2004.00   Feminino                    1396298.00   
2 2004.00  Masculino                    1959787.00   
3 2005.00   Feminino                    1460602.00   
5 2005.00  Masculino                    2041717.00   

   Valor Benefícios Ativos (R$)  
0                    3443513.27  
2                    7074641.34  
3                    3937167.26  
5                    8124307.75

In [122]:
for idx, df_grupo in df[['Ano', 'Sexo', 'Quantidade Benefícios Ativos', 'Valor Benefícios Ativos (R$)']]\
    .groupby(['Ano', 'Sexo'], as_index=False):
    display(idx)
    display(df_grupo)
    break

(2004.0, 'Feminino')

Ano      Sexo  Quantidade Benefícios Ativos  \
1   2004.00  Feminino                       9765.00   
4   2004.00  Feminino                      12185.00   
7   2004.00  Feminino                       5348.00   
10  2004.00  Feminino                        688.00   
13  2004.00  Feminino                      78846.00   
16  2004.00  Feminino                      48579.00   
19  2004.00  Feminino                     130493.00   
22  2004.00  Feminino                       5816.00   
25  2004.00  Feminino                       1493.00   
28  2004.00  Feminino                       2533.00   
31  2004.00  Feminino                         31.00   
34  2004.00  Feminino                       3152.00   
37  2004.00  Feminino                      84484.00   
40  2004.00  Feminino                        403.00   
43  2004.00  Feminino                          7.00   
46  2004.00  Feminino                       1780.00   
49  2004.00  Feminino                      14872.00   
52  2004.00  Feminino                        130.00   
55  2004.00  Feminino                       2330.00   
58  2004.00  Feminino                       8489.00   
61  2004.00  Feminino                      16465.00   
64  2004.00  Feminino                        659.00   
67  2004.00  Feminino                       7773.00   
70  2004.00  Feminino                      37836.00   
73  2004.00  Feminino                      15915.00   
76  2004.00  Feminino                      12189.00   
79  2004.00  Feminino                        982.00   
82  2004.00  Feminino                      84435.00   
85  2004.00  Feminino                       6172.00   
88  2004.00  Feminino                       1864.00   
..      ...       ...                           ...   
391 2004.00  Feminino                          0.00   
394 2004.00  Feminino                          0.00   
397 2004.00  Feminino                         13.00   
400 2004.00  Feminino                          0.00   
403 2004.00  Feminino                          0.00   
406 2004.00  Feminino                          0.00   
409 2004.00  Feminino                         14.00   
412 2004.00  Feminino                          0.00   
415 2004.00  Feminino                          1.00   
418 2004.00  Feminino                         29.00   
421 2004.00  Feminino                         50.00   
424 2004.00  Feminino                         13.00   
427 2004.00  Feminino                         15.00   
430 2004.00  Feminino                        575.00   
433 2004.00  Feminino                       3174.00   
436 2004.00  Feminino                         63.00   
439 2004.00  Feminino                         57.00   
442 2004.00  Feminino                        212.00   
445 2004.00  Feminino                        232.00   
448 2004.00  Feminino                          6.00   
451 2004.00  Feminino                        100.00   
454 2004.00  Feminino                      19802.00   
457 2004.00  Feminino                         14.00   
460 2004.00  Feminino                          4.00   
463 2004.00  Feminino                          2.00   
466 2004.00  Feminino                          0.00   
469 2004.00  Feminino                       8743.00   
472 2004.00  Feminino                          0.00   
475 2004.00  Feminino                         13.00   
478 2004.00  Feminino                         43.00   

     Valor Benefícios Ativos (R$)  
1                      2343602.54  
4                      3661307.37  
7                      1623786.64  
10                      193678.34  
13                    21194847.19  
16                    13206860.59  
19                    32231396.57  
22                     1470580.44  
25                      432598.47  
28                      780805.96  
31                        8857.21  
34                      768688.46  
37                    21938691.64  
40                      122684.17  
43                        2071.91  
46                   

In [124]:
# Agrupamento por mais de uma coluna e uso de agregações específicas para cada coluna
agregacoes = {
    'Quantidade Benefícios Ativos':{
        'Média/CID': 'mean',
        'Total': 'sum',
    },
    'Valor Benefícios Ativos (R$)':{
        'Média/CID': 'mean',
        'Total': 'sum',
    }
}

df_ano_sexo = df[
    ['Ano', 'Sexo', 'Quantidade Benefícios Ativos', 'Valor Benefícios Ativos (R$)']]\
    .groupby(['Ano', 'Sexo'], as_index=False)\
    .aggregate(agregacoes)
df_ano_sexo.head(6)

Ano       Sexo Quantidade Benefícios Ativos             \
                                        Média/CID      Total   
0 2004.00   Feminino                      8726.86 1396298.00   
1 2004.00   Ignorado                       182.34   29175.00   
2 2004.00  Masculino                     12248.67 1959787.00   
3 2005.00   Feminino                      9128.76 1460602.00   
4 2005.00   Ignorado                       212.58   25510.00   
5 2005.00  Masculino                     12760.73 2041717.00   

  Valor Benefícios Ativos (R$)                
                     Média/CID         Total  
0                   3443513.27  550962123.88  
1                     48180.35    7708856.49  
2                   7074641.34 1131942614.36  
3                   3937167.26  629946761.14  
4                     59876.16    7185139.31  
5                   8124307.75 1299889240.05

# Concat

Utilizado para incluir as linhas ou colunas de um dataframe em outro.

In [137]:
np.random.seed(42)

df1 = pd.DataFrame({
    'A': list(np.random.randint(1, high=100, size=3)),
    'B': list(np.random.randint(1, high=100, size=3)),
})

df2 = pd.DataFrame({
    'A': list(np.random.randint(1, high=100, size=4)),
    'B': list(np.random.randint(1, high=100, size=4)),
})

In [138]:
df1

A   B
0  52  72
1  93  61
2  15  21

In [139]:
df2

A   B
0  83  88
1  87  24
2  75   3
3  75  22

In [148]:
pd.concat([df1, df2, pd.DataFrame([[52,72]], columns=['A', 'B'], index=[23,24])], ignore_index=True, axis=1)

0     1     2     3     4     5
0  52.00 72.00 83.00 88.00   nan   nan
1  93.00 61.00 87.00 24.00   nan   nan
2  15.00 21.00 75.00  3.00   nan   nan
3    nan   nan 75.00 22.00   nan   nan
23   nan   nan   nan   nan 52.00 72.00
24   nan   nan   nan   nan 52.00 72.00

In [29]:
pd.concat([df1, df2], ignore_index=False, axis=1)

A     B   A   B
0 35.00 19.00  75   2
1 12.00 63.00  53  93
2 55.00 61.00  43  81
3   nan   nan  83  16

# Merge

In [150]:
df_pericias = pd.read_csv('dados/PMD04.csv', sep=',', encoding='latin1')
df_pericias_credenciadas = df_pericias[df_pericias['Especialidade'].str.contains('Perí')]
df_pericias_credenciadas.head()

Ano                    Especialidade  Qtde Exames Complementares  \
404  2011  Perícias Médicas (Credenciadas)                      162481   
439  2012  Perícias Médicas (Credenciadas)                           0   
440  2013  Perícias Médicas (Credenciadas)                       28097   
458  2014  Perícias Médicas (Credenciadas)                       49171   
476  2015  Perícias Médicas (Credenciadas)                       43425   

    Vlr Exames Complementares (R$)  
404                     3132612,00  
439                           0,00  
440                      960041,00  
458                     1720977,00  
476                     1519875,00

In [158]:
df_ano = df[['Ano', 'Quantidade Benefícios Ativos']].groupby(['Ano'], as_index=False).sum()
df_ano.head()

Ano  Quantidade Benefícios Ativos
0 2004.00                    3385260.00
1 2005.00                    3527829.00
2 2006.00                    3694521.00
3 2007.00                    3573442.00
4 2008.00                    3624670.00

In [159]:
df_merge = pd.merge(df_ano, df_pericias_credenciadas[['Ano','Especialidade']], how='inner', left_on='Ano', right_on='Ano')
df_merge.head()

Ano  Quantidade Benefícios Ativos                    Especialidade
0 2011.00                    4538149.00  Perícias Médicas (Credenciadas)
1 2012.00                    4635979.00  Perícias Médicas (Credenciadas)
2 2013.00                    4828086.00  Perícias Médicas (Credenciadas)
3 2014.00                    4975810.00  Perícias Médicas (Credenciadas)
4 2015.00                    4966702.00  Perícias Médicas (Credenciadas)

In [156]:
df_merge = pd.merge(df_ano, df_pericias_credenciadas, how='outer', left_on='Ano', right_on='Ano', indicator=True)
df_merge.head(10)

Ano  Quantidade Benefícios Ativos                    Especialidade  \
0 2004.00                    3385260.00                              NaN   
1 2005.00                    3527829.00                              NaN   
2 2006.00                    3694521.00                              NaN   
3 2007.00                    3573442.00                              NaN   
4 2008.00                    3624670.00                              NaN   
5 2009.00                    3576679.00                              NaN   
6 2010.00                    4442648.00                              NaN   
7 2011.00                    4538149.00  Perícias Médicas (Credenciadas)   
8 2012.00                    4635979.00  Perícias Médicas (Credenciadas)   
9 2013.00                    4828086.00  Perícias Médicas (Credenciadas)   

   Qtde Exames Complementares Vlr Exames Complementares (R$)     _merge  
0                         nan                            NaN  left_only  
1                         nan                            NaN  left_only  
2                         nan                            NaN  left_only  
3                         nan                            NaN  left_only  
4                         nan                            NaN  left_only  
5                         nan                            NaN  left_only  
6                         nan                            NaN  left_only  
7                   162481.00                     3132612,00       both  
8                        0.00                           0,00       both  
9                    28097.00                      960041,00       both

# Pivot Table

In [160]:
df_ano_sexo.head()

Ano       Sexo Quantidade Benefícios Ativos             \
                                        Média/CID      Total   
0 2004.00   Feminino                      8726.86 1396298.00   
1 2004.00   Ignorado                       182.34   29175.00   
2 2004.00  Masculino                     12248.67 1959787.00   
3 2005.00   Feminino                      9128.76 1460602.00   
4 2005.00   Ignorado                       212.58   25510.00   

  Valor Benefícios Ativos (R$)                
                     Média/CID         Total  
0                   3443513.27  550962123.88  
1                     48180.35    7708856.49  
2                   7074641.34 1131942614.36  
3                   3937167.26  629946761.14  
4                     59876.16    7185139.31

In [161]:
pd.pivot_table(df_ano_sexo, index='Ano', columns='Sexo')

Quantidade Benefícios Ativos                                         \
                           Média/CID                         Total            
Sexo                        Feminino Ignorado Masculino   Feminino Ignorado   
Ano                                                                           
2004.00                      8726.86   182.34  12248.67 1396298.00 29175.00   
2005.00                      9128.76   212.58  12760.73 1460602.00 25510.00   
2006.00                       162.97   163.43    185.99 1528213.00 22553.00   
2007.00                        64.46   122.10     71.25 1461801.00 19903.00   
2008.00                        66.97     8.82     74.53 1494278.00  3536.00   
2009.00                        67.70     7.78     74.82 1480592.00  3012.00   
2010.00                        80.62     7.81     82.28 1805323.00 16456.00   
2011.00                        83.49     6.98     85.17 1857564.00 13062.00   
2012.00                        86.90     6.54     88.45 1909594.00 11407.00   
2013.00                        92.77     6.12     93.49 2008280.00  9846.00   
2014.00                        97.95     5.68     97.96 2087519.00  8559.00   
2015.00                       100.76     5.38     99.93 2095713.00  7620.00   
2016.00                       104.24     4.97    103.85 2122685.00  6367.00   

                   Valor Benefícios Ativos (R$)                      \
                                      Média/CID                       
Sexo     Masculino                     Feminino Ignorado  Masculino   
Ano                                                                   
2004.00 1959787.00                   3443513.27 48180.35 7074641.34   
2005.00 2041717.00                   3937167.26 59876.16 8124307.75   
2006.00 2143755.00                     75864.13 48308.95  127508.52   
2007.00 2091738.00                     31182.93 37458.39   50891.53   
2008.00 2126856.00                     34554.57  2903.68   56696.30   
2009.00 2093075.00                     37165.55  2750.05   60920.50   
2010.00 2620869.00                     54165.76  4056.34   72610.78   
2011.00 2667523.00                     59983.30  3879.30   80679.28   
2012.00 2714978.00                     68917.41  4146.00   91120.93   
2013.00 2809960.00                     79945.01  4223.12  104722.94   
2014.00 2879732.00                     89948.29  4180.95  116775.42   
2015.00 2863369.00                     99596.26  4306.53  127828.57   
2016.00 2889412.00                    114766.93  4449.22  148105.02   

                                                
                Total                           
Sexo         Feminino   Ignorado     Masculino  
Ano                                             
2004.00  550962123.88 7708856.49 1131942614.36  
2005.00  629946761.14 7185139.31 1299889240.05  
2006.00  711377942.46 6666634.55 1469663190.91  
2007.00  707197617.02 6105717.53 1494073565.10  
2008.00  770981564.48 1164373.89 1617828911.91  
2009.00  812773501.74 1064269.42 1704311972.81  
2010.00 1212933871.90 8542652.36 2312943671.60  
2011.00 1334508538.88 7258177.08 2526794264.81  
2012.00 1514460105.55 7226469.99 2796865926.42  
2013.00 1730649505.71 6795006.38 3147447999.11  
2014.00 1917067874.37 6304873.51 3432846944.00  
2015.00 2071602281.85 6098040.62 3662671987.30  
2016.00 2337113757.98 5699455.60 4120726031.85

# Melt

In [36]:
calendario = {'Dia da Semana': ["Segunda", "Terça", "Quarta", "Quinta", "Sexta", "Sábado", "Domingo"],
        'Operador 1': [14, 6, 8, 4, 13, 6, 4],
        'Operador 2': [12, 4, 17, 8, 9, 12, 5],
        'Operador 3': [4, 2, 7, 6, 7, 9, 13]}
df_calendario = pd.DataFrame(calendario)
df_calendario.head()

Dia da Semana  Operador 1  Operador 2  Operador 3
0       Segunda          14          12           4
1         Terça           6           4           2
2        Quarta           8          17           7
3        Quinta           4           8           6
4         Sexta          13           9           7

In [37]:
df_melt = pd.melt(df_calendario, id_vars=["Dia da Semana"], var_name="Operador", value_name="Atendimentos")
df_melt.head()

Dia da Semana    Operador  Atendimentos
0       Segunda  Operador 1            14
1         Terça  Operador 1             6
2        Quarta  Operador 1             8
3        Quinta  Operador 1             4
4         Sexta  Operador 1            13

# Preenchimento de Valores Faltantes

In [163]:
"""
O que o pandas considera como valores faltantes?
"""

# padrão do pandas é False
pd.options.mode.use_inf_as_na = False 
a = pd.Series([np.NaN,1,2,3,-5, None])
b = pd.Series([3,2,1,0,0, 1])
c = a / b
c

0    nan
1   0.50
2   2.00
3    inf
4   -inf
5    nan
dtype: float64

In [167]:
np.inf - np.inf

nan

In [164]:
c.isnull()

0     True
1    False
2    False
3    False
4    False
5     True
dtype: bool

In [168]:
pd.options.mode.use_inf_as_na = True
c.isnull()

0     True
1    False
2    False
3     True
4     True
5     True
dtype: bool

In [169]:
lista = [1,np.nan,np.nan,np.nan,5,np.nan,7,np.nan,9,np.nan,11]

pd.DataFrame({
    'Original': pd.Series(lista),
    'fillna': pd.Series(lista).fillna(0),
    'ffill': pd.Series(lista).ffill(limit=1),
    'bfill': pd.Series(lista).bfill(limit=1),    
    'interpolate': pd.Series(lista).interpolate(),
})   
    
    

Original  fillna  ffill  bfill  interpolate
0       1.00    1.00   1.00   1.00         1.00
1        nan    0.00   1.00    nan         2.00
2        nan    0.00    nan    nan         3.00
3        nan    0.00    nan   5.00         4.00
4       5.00    5.00   5.00   5.00         5.00
5        nan    0.00   5.00   7.00         6.00
6       7.00    7.00   7.00   7.00         7.00
7        nan    0.00   7.00   9.00         8.00
8       9.00    9.00   9.00   9.00         9.00
9        nan    0.00   9.00  11.00        10.00
10     11.00   11.00  11.00  11.00        11.00

In [170]:
lista = [10,np.nan,np.nan,np.nan,5,np.nan,7,np.nan,3,np.nan,11]

df = pd.DataFrame({
    'Original': pd.Series(lista),
    'fillna': pd.Series(lista).fillna(0),
    'ffill': pd.Series(lista).ffill(limit=1),
    'bfill': pd.Series(lista).bfill(limit=1),    
    'interpolate': pd.Series(lista).interpolate(),
})  

## Importação de Dados com o Pandas

O pandas possui diversas funcionalidades auxiliares para importar dados de diversas fontes de maneira simples. Essas funcionalidades podem ser combinadas com outras bibliotecas e funções do python para acessarmos praticamente qualquer fonte de dados.

Vamos ver alguns casos bastante comuns sem a pretensão de esgotar o assunto.

### Arquivos delimitados

Arquivos delimitados podem ser lidos e transforados em DataFrames utilizando a função __pd.read_csv()__. Entretanto, estes aqrquivo possuem algumas características próprias. Geralmente temos um caracter separador de colunas e cada linha representa um registro. Ainda assim, há casos em que o separador de linhas não é uma quebra de linha. Há também casos em que números e datas não estão no formato correto para que o pandas faça o conversão para o tipo de dados correto.

Assim, descrevemos parâmetros desta função que ajudam a resolver os principais problemas. 

* __encoding__: cada arquivo texto possui um encoding. O encoding define o formato utilizado para representar os caracteres de um arquivo texto. O encoding padrão do python é 'utf8'. Entretanto, é comum ver arquivos em encodings diferentes para representar caracteres da língua portuguesa. Os mais comuns são: 'iso-8859-1', 'latin1' e 'cp1252'. O principal problema de encoding errado é percebido na forma como os caracteres acentuadas são exibidos. Mais detalhes em: https://docs.python.org/3.6/library/codecs.html#standard-encodings

* __thousands__ e __decimal__: são utilizados para informar em qual formato os dados numéricos estão representados no arquivo. Utiliza-se __thousands__ para informar o separador de milhar e __decimal__ para informar o separador de milhares

* __parse_dates__ e __dayfirst__: utilizados para informar se datas devem ser convertidas para o formato de datas do python.

* __iterator__ e __chunksize__: utilizados para definiar um carregamento em blocos para casos onde os dados são muito maiores do que a memória.

In [171]:
# exemplo completo
arquivo = 'dados/exemplo_delimitado.zip'
df_exemplo_delimitado = pd.read_csv(
    arquivo, 
    sep=';', 
    thousands='.', 
    decimal=',', 
    parse_dates=True, 
    dayfirst=True, 
    index_col='Data',
    )
df_exemplo_delimitado.head()

Valor
Data               
2019-02-13 80299.81
2019-08-30 30787.57
2019-05-26 82908.59
2018-07-02 75336.25
2019-06-21 16934.76

In [172]:
# consulta fatiando pelo índice
df_exemplo_delimitado.loc['2018-06':'2018-07'].head()

Valor
Data               
2018-07-02 75336.25
2018-06-03 10524.73
2018-07-26 51842.44
2018-06-25  7190.29
2018-07-20 17242.60

In [173]:
# detalhes sobre os tipos convertidos
df_exemplo_delimitado.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 100000 entries, 2019-02-13 to 2018-11-01
Data columns (total 1 columns):
Valor    100000 non-null float64
dtypes: float64(1)
memory usage: 4.0 MB


In [174]:
# reorganização dos dados por mês, com a soma mensal
df_exemplo_delimitado.resample('1M').sum().sort_index().head()

Valor
Data                   
2018-05-31 303285672.99
2018-06-30 305561462.09
2018-07-31 310799198.36
2018-08-31 307770016.93
2018-09-30 292280527.94

In [175]:
# Exemplo de leitura de um arquivo de 100K registros 
# em blocos de 10K.
reader = pd.read_csv(
    arquivo, 
    sep=';', 
    thousands='.', 
    decimal=',', 
    parse_dates=True, 
    dayfirst=True, 
    index_col='Data',
    chunksize=10000,
    iterator=True,
    )

In [176]:
type(reader)

pandas.io.parsers.TextFileReader

In [177]:
somas = 0
for idx, df_parcial in enumerate(reader):
    somas += df_parcial['Valor'].sum()
    print("Bloco {}, Soma parcial: {:0.2f}".format(idx+1, somas))

Bloco 1, Soma parcial: 502357350.05
Bloco 2, Soma parcial: 1005129782.91
Bloco 3, Soma parcial: 1508380506.49
Bloco 4, Soma parcial: 2009881810.00
Bloco 5, Soma parcial: 2514613689.05
Bloco 6, Soma parcial: 3015215014.32
Bloco 7, Soma parcial: 3515577625.84
Bloco 8, Soma parcial: 4015867740.46
Bloco 9, Soma parcial: 4519566546.25
Bloco 10, Soma parcial: 5017208156.37


### Arquivos Excel

O pandas também possui uma função espcífica para importar dados de arquivos excel.

In [178]:
# leitura de um arquivo excel
df_exemplo_excel = pd.read_excel('dados/exemplo_delimitado.xlsx')
df_exemplo_excel.head()

Data    Valor               Descrição
0 2019-02-13 80299.81  Uma descrição qualquer
1 2019-08-30 30787.57  Uma descrição qualquer
2 2019-05-26 82908.59  Uma descrição qualquer
3 2018-07-02 75336.25  Uma descrição qualquer
4 2019-06-21 16934.76  Uma descrição qualquer

In [179]:
df_exemplo_excel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99 entries, 0 to 98
Data columns (total 3 columns):
Data         99 non-null datetime64[ns]
Valor        99 non-null float64
Descrição    99 non-null object
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 2.4+ KB


### Bancos de Dados

Existem várias formas de conectar em Sistemas de Bancos de Dados utilizando python e então converter os resultados em DataFrames. Abordaremos aqui umas das formas mais simples.

Utilizaremos uma biblioteca chamada sqlalchemy.


Exemplos de conexões com sqlalchemy para os principais bancos de dados podem ser encontrados aqui: http://docs.sqlalchemy.org/en/latest/core/engines.html.

Além do sqlalchemy é necessário ter a biblioteca (também chamada de conector ou driver) do banco de dados que será utilizado.

In [180]:
from sqlalchemy import create_engine
arquivo = os.path.join('arquivos','exemplo_sqlite.db')

engine = create_engine('sqlite:///dados/exemplo_sqlite.db')

query = "select * from tb_exemplo_sqlite"
df_sqlite = pd.read_sql(sql=query, con=engine)
df_sqlite.head()

Data   Valor              Descrição
0  07/11/2018 1728.49  Decrição produto 4157
1  12/06/2018 9607.77  Decrição produto 5831
2  03/10/2018 1197.53  Decrição produto 6348
3  12/06/2018 9172.97  Decrição produto 6600
4  19/08/2019 1845.59  Decrição produto 6570

O exemplo acima funcionaria para um banco SQLServer, por exemplo, fazendo os seguites passos:

1. Instalar o sqlalchemy e o pyodbc
```bash
pip install sqlalchemy pyodbc
```

2. Criar uma conexão odbc com o banco de dados na máquina onde o jupyter está rodando e dar um nome para ela. Ex: DSN_TESTE

3. Fazer uma conexão utlizando o sqlalchemy.
```python
engine = create_engine('mssql+pyodbc://<usuario>:<senha>@DSN_TESTE')
df_mssql = pd.read_sql(sql=query, con=engine)
df_mssql.head()
```



#### #python 
* para mais detalhes sobre o sqlalchemy veja http://docs.sqlalchemy.org/en/latest/index.html

### Parquet

Formato de armazenamento baseado em colunas que armazena os dados de forma compactada além de preservar os metadas dos tipos de cada coluna. Bastante utilizado no ecosistema hadoop.

In [51]:
!pip install pyarrow

In [182]:
%%time
df_parquet = pd.read_parquet('dados/arq_municipios_fronteiricos.parquet')
df_parquet.head()

Wall time: 6.99 ms


Município             Estado  Área territorial  \
0                 Aceguá  Rio Grande do Sul           1550.00   
1             Acrelândia               Acre           1575.00   
2                Alecrim  Rio Grande do Sul            315.00   
3               Almeirim               Pará          72960.00   
4  Alta Floresta d'Oeste           Rondônia           7067.00   

   População (IBGE/2007)  Densidade demográfica (hab/km2)  PIB (IBGE/2005)  \
0                4138.00                             2.66      71638000.00   
1               11520.00                             7.31     114350000.00   
2                7357.00                            23.35      44373000.00   
3               30903.00                             0.42     462258000.00   
4               23857.00                             3.37     186812000.00   

   PIB per capita (R$)  IDH/2000  
0             17266.00       nan  
1              9986.00      0.68  
2              5944.00      0.74  
3             13485.00      0.74  
4              6525.00      0.71

In [53]:
%%javascript

(function() {var css = [
    ".container {width: 100%;}",
    ".text_cell_render table {font-size: 24;}",
    ".rendered_html code {background-color: #fafafa;}",
    ".rendered_html :not(pre) > code {padding: 1px 1px;}",
    ".rendered_html pre code {background-color: #fafafa;}",
    ".rendered_html pre {border: 1px;background-color: #fafafa;padding: .5ex .5em;}"
].join("\n");
if (typeof GM_addStyle != "undefined") {
    GM_addStyle(css);
} else if (typeof PRO_addStyle != "undefined") {
    PRO_addStyle(css);
} else if (typeof addStyle != "undefined") {
    addStyle(css);
} else {
    var node = document.createElement("style");
    node.type = "text/css";
    node.appendChild(document.createTextNode(css));
    var heads = document.getElementsByTagName("head");
    if (heads.length > 0) {
        heads[0].appendChild(node);
    } else {
        document.documentElement.appendChild(node);
    }
}
})();

<IPython.core.display.Javascript object>

___
__Material produzido para o curso__:
* Introdução à Análise de Dados com Python

__Autor__:
* Fernando Sola Pereira

__Revisão__:
* 1.1